braindecode/preprocessing/preprocess.py

```diff
@@ -120,54 +126,114 @@ def preprocess(concat_ds, preprocessors):
         A concat of BaseDataset or WindowsDataset datasets to be preprocessed.
     preprocessors: list(Preprocessor)
         List of Preprocessor objects to apply to the dataset.
+    save_dir : str | None
+        If a string, the preprocessed data will be saved under the specified
+        directory and the datasets in ``concat_ds`` will be reloaded with
+        `preload=False`.
+    overwrite : bool
+        When `save_dir` is provided, controls whether to overwrite existing
+        files in `save_dir`.
+    n_jobs : int | None
+        Number of jobs for parallel execution.
 
     Returns
     -------
     BaseConcatDataset:
         Preprocessed dataset.
     """
+    # In case of serialization, make sure directory is available before
+    # preprocessing
+    if save_dir is not None and not overwrite:
+        _check_save_dir_empty(save_dir)
+
     if not isinstance(preprocessors, Iterable):
         raise ValueError(
             'preprocessors must be a list of Preprocessor objects.')
     for elem in preprocessors:
         assert hasattr(elem, 'apply'), (
             'Preprocessor object needs an `apply` method.')
 
-    # get the preprocessing keyword arguments
-    preproc_kwargs = _get_preproc_kwargs(preprocessors)
+    list_of_ds = Parallel(n_jobs=n_jobs)(
+        delayed(_preprocess)(ds, i, preprocessors, save_dir, overwrite)
+        for i, ds in enumerate(concat_ds.datasets))
 
-    for ds in concat_ds.datasets:
-        if hasattr(ds, 'raw'):
-            _preprocess(ds.raw, preprocessors)
-            # store raw preprocessing keyword arguments to the dataset
-            setattr(ds, 'raw_preproc_kwargs', preproc_kwargs)
-        elif hasattr(ds, 'windows'):
-            _preprocess(ds.windows, preprocessors)
-            # store window preprocessing keyword arguments to the dataset
-            setattr(ds, 'window_preproc_kwargs', preproc_kwargs)
-        else:
-            raise ValueError(
-                'Can only preprocess concatenation of BaseDataset or '
-                'WindowsDataset, with either a `raw` or `windows` attribute.')
+    if save_dir is not None:  # Reload datasets and replace in concat_ds
+        concat_ds_reloaded = load_concat_dataset(
+            save_dir, preload=False, target_name=None)
+        _replace_inplace(concat_ds, concat_ds_reloaded)
+    else:
+        if n_jobs is None or n_jobs == 1:  # joblib did not make copies, the
+            # preprocessing happened in-place
+            # Recompute cumulative sizes as transforms might have changed them
+            concat_ds.cumulative_sizes = concat_ds.cumsum(concat_ds.datasets)
+        else:  # joblib made copies
+            _replace_inplace(concat_ds, BaseConcatDataset(list_of_ds))
+
+    return concat_ds
+
+
+def _replace_inplace(concat_ds, new_concat_ds):
+    """Inplace modification of a BaseConcatDataset with another's attributes.
```

Comment:
Could you write more here, like all "subdatasets" will be replaced and also the preproc_kwargs